azuredevops/internal/service/permissions/resource_build_definition_permissions.go

```diff
@@ -0,0 +1,170 @@
+		Update: resourceBuildDefinitionPermissionsCreateOrUpdate,
+		Delete: resourceBuildDefinitionPermissionsDelete,
+		Schema: securityhelper.CreatePermissionResourceSchema(map[string]*schema.Schema{
+			"project_id": {
+				Type:         schema.TypeString,
+				ValidateFunc: validation.IsUUID,
+				Required:     true,
+				ForceNew:     true,
+			},
+			"build_definition_id": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"build_definition_path": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+		}),
+	}
+}
+
+func resourceBuildDefinitionPermissionsCreateOrUpdate(d *schema.ResourceData, m interface{}) error {
+	clients := m.(*client.AggregatedClient)
+
+	sn, err := securityhelper.NewSecurityNamespace(d, clients, securityhelper.SecurityNamespaceIDValues.Build, createBuildToken)
+	if err != nil {
+		return err
+	}
+
+	if err := securityhelper.SetPrincipalPermissions(d, sn, nil, false); err != nil {
+		return err
+	}
+
+	return resourceBuildDefinitionPermissionsRead(d, m)
+}
+
+func resourceBuildDefinitionPermissionsRead(d *schema.ResourceData, m interface{}) error {
+	clients := m.(*client.AggregatedClient)
+
+	sn, err := securityhelper.NewSecurityNamespace(d, clients, securityhelper.SecurityNamespaceIDValues.Build, createBuildToken)
+	if err != nil {
+		return err
+	}
+
+	principalPermissions, err := securityhelper.GetPrincipalPermissions(d, sn)
+	if err != nil {
+		return err
+	}
+	if principalPermissions == nil {
+		d.SetId("")
+		log.Printf("[INFO] Permissions for ACL token %q not found. Removing from state", sn.GetToken())
+		return nil
+	}
+
+	d.Set("permissions", principalPermissions.Permissions)
+	return nil
+}
+
+func resourceBuildDefinitionPermissionsDelete(d *schema.ResourceData, m interface{}) error {
+	clients := m.(*client.AggregatedClient)
+
+	sn, err := securityhelper.NewSecurityNamespace(d, clients, securityhelper.SecurityNamespaceIDValues.Build, createBuildToken)
+	if err != nil {
+		return err
+	}
+
+	if err := securityhelper.SetPrincipalPermissions(d, sn, &securityhelper.PermissionTypeValues.NotSet, true); err != nil {
+		return err
+	}
+	d.SetId("")
+	return nil
+}
+
+func createBuildToken(d *schema.ResourceData, clients *client.AggregatedClient) (string, error) {
+	projectID, ok := d.GetOk("project_id")
+	if !ok {
+		return "", fmt.Errorf("Failed to get 'project_id' from schema")
+	}
+
+	buildDefinitionID, err := getBuildDefinitionID(d)
+	if err != nil {
+		return "", err
+	}
+
+	definition, err := clients.BuildClient.GetDefinition(clients.Ctx, build.GetDefinitionArgs{
+		Project:      converter.String(projectID.(string)),
+		DefinitionId: converter.Int(buildDefinitionID),
+	})
+
+	if err != nil {
+		return "", err
+	}
+
+	var aclToken string
+
+	// The token format is Project_ID/Build_Definition_ID
+	// or Project_ID/Path/Build_Definition_ID
+
+	if *definition.Path != "\\" {
+		path, ok := d.GetOk("build_definition_path")
```

Comment:
Since we have know the project ID and definition ID, suggest get the build path from service:
```go
definition, err := clients.BuildClient.GetDefinition(clients.Ctx, build.GetDefinitionArgs{
		Project:      converter.String(projectID.(string)),
		DefinitionId: converter.Int(buildDefinitionID),
})
```

Reply:
Hi @xuzhang3 , I had this in my original implementation.
However when I tested changing the path, I had to apply twice for the `azuredevops_build_definition_permissions` to be updated.

On the first apply, only the `azuredevops_build_definition` would change due to the path variable.
However, the `azuredevops_build_definition_permissions` did not changed.

On the 2nd apply, `azuredevops_build_definition` did not change.
However, the `azuredevops_build_definition_permissions` now reads the new path, and is re-created.

I opted for the additional parameter - `build_definition_path` - so that the change would be completed in a single terraform apply.

How would you like to proceed?

Reply:
Hi @josh-barker-coles I remove the `build_definition_path` configuration. Permissions and path can be changed at the same time. 
My chagnes:
```go
func createBuildToken(d *schema.ResourceData, clients *client.AggregatedClient) (string, error) {
	projectID, ok := d.GetOk("project_id")
	if !ok {
		return "", fmt.Errorf("Failed to get 'project_id' from schema")
	}

	buildDefinitionID, err := getBuildDefinitionID(d)
	if err != nil {
		return "", err
	}

	definition, err := clients.BuildClient.GetDefinition(clients.Ctx, build.GetDefinitionArgs{
		Project:      converter.String(projectID.(string)),
		DefinitionId: converter.Int(buildDefinitionID),
	})

	if err != nil {
		return "", err
	}

	var aclToken string

	// The token format is Project_ID/Build_Definition_ID
	// or Project_ID/Path/Build_Definition_ID

	if *definition.Path != "\\" {
		//path := d.GetOk("build_definition_path")
		//if !ok {
		//	return "", fmt.Errorf("Set path to %s", *definition.Path)
		//}

		//if path.(string) != *definition.Path {
		//	return "", fmt.Errorf("Update path to %s", *definition.Path)
		//}

		transformedPath := transformPath(*definition.Path)

		aclToken = fmt.Sprintf("%s/%s/%d", projectID.(string), transformedPath, buildDefinitionID)
	} else {
		aclToken = fmt.Sprintf("%s/%d", projectID.(string), buildDefinitionID)
	}

	return aclToken, nil
}
```

Reply:
I've tested again and I got the same result.

I've updated the code, and updated the acceptance tests.